rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+
+    * - WRITING
+      - The Client is writing a new firmware image to the *second*, in preparation for installation.
+
+        When writing is complete, it can be installed.
+
+    * - STAGED
+      - Installation of the *second* has been requested, but the system must be restarted as the final update operation runs within the Bootloader.
+
+        This state is transient.
+
+    * - FAILED
+      - An installation of the *second* has been attempted, but has failed for some reason. The failure reason is recorded in the Store.
+
+        The *second* needs to be erased before another update can be attempted.
+
+    * - TRIAL
+      - Installation of the *second* has succeeded, and is now the *active* running in 'trial mode'. This state is transient, and requires the Client to explicitly accept the trial to make the update permanent.
+
+        In this state, the previously installed *active* image is preserved as the *second*. If the trial is explicitly rejected, or the system restarts without accepting the trial, the previously installed image is re-installed and the trial image is rejected.
+
+    * - REJECTED
+      - The *active* trial image has been rejected, but the system must be restarted so the bootloader can revert to the previous image (saved as the *second*).
+
+        This state is transient.
+
+    * - UPDATED
+      - The *active* trial image has been accepted, and is now permanently active.
+
+        The *second* contains the now-expired previous firmware image, which needs to be erased before another update can be started.
+
+Note
+    In the basic flow here, REJECTED seems to have limited difference from TRIAL, other than that it records that the trial has been explicitly rejected.
+
+    However, it provides two important capabilities:
+
+    * It specifically prevents the Client accepting a trial after rejecting it
+    * It enables other TRIAL policies, such as permitting a limited number of restarts (but more than one) before automatically reverting to the previous image.
+
+State transitions
+~~~~~~~~~~~~~~~~~
+
+The state transitions occur either as a result of an API call from the Client, or when the Bootloader carries out an installation operation. The installation operations that occur when the Bootloader runs are determined by the state of the Firmware Store, and do not depend on the reason for the restart. To ensure a specific action occurs at restart, the Client uses the appropriate API to set the state of the Firmware Store.
+
+The Client can trigger transitions in the state model using the following operations:
+
+* ``start``
+* ``write``
+* ``install``
+* ``erase``
+* ``accept``
+* ``reject``
+
+Some transitions can only be triggered by a system restart, which will run the Bootloader. For these transitions it does not matter how the reboot was triggered.
+
+The typical flow through the states is shown in this graphic:
+
+.. image:: fwu-states-simple.svg
+
```

Comment:
Is state swich from  STAGED to TRIVIAL and then to UPDATED a mandatory way for a successful image update? Some bootloaders may not support TRIVIAL state: if the image in the secondary slot is verified and it's version is newer than that in the active slot, then the image in the secondary slot is accepted as a permanent image.

Reply:
That is one of the questions we would like a response to in the Scope discussion (#2), and survey. The state model here includes TRIAL state, because it is more complex than the state model for a component that does not require TRIAL.

My expectation is that the final specification will also provide the subset of the full state model for the in-scope component designs. In this case, the state model would just remove the TRIAL and REJECTED states, and result in the following flow:

![image](https://user-images.githubusercontent.com/5690199/169814588-1b7963d7-dd3f-487c-bcfa-cb5f7983c268.png)

If the firmware is subject to a rollback-prevention policy, then a transition from STAGED to UPDATED can brick the device if the updated firmware does not work. If the firmware is well enough tested before deployment, then this is perhaps an acceptable trade-off.

Reply:
Isn't TRIAL state, at the minimum, about the firmware calling accept? So the, possibly very brief, time between firmware calling install() and accept() is spent in TRIAL state?

Reply:
If the system is not booted into TRIAL state (to the point that the Client can see the state), then for purpose of this API, the state does not exist.

The bootloader might internally mark the image as TRIAL during the update process, but accept the image prior to starting the application firmware (and the Update Client).

Reply:
Agree. The psa_fwu_accept is not needed if the bootloader does not support a TRIAL state. If it becomes brick, then it is not the FWU or OTA process that should be blamed. It is the issue of the bootloader itself. I think it makes sense to either forbid FWU API using on no TRIVIAL state bootloaders and make the TRIVIAL state mandatory or remind the brick risk and make the TRIVIAL state optional.